File: src/lib/gis-api.ts
/**
 * Typed client for the GIS Environmental Boundary Check API.
 *
 * Every call reaches the real FastAPI service (via the same-origin proxy at
 * /api/gis/*, which attaches the server-held bearer token). Nothing here is mocked or
 * stubbed, and no response value is synthesised on the client -- if the backend is down
 * or unauthorised, that surfaces as a typed error rather than as placeholder data.
 *
 * Failures are modelled as distinct classes so the UI can render the right state
 * instead of collapsing everything into one "something went wrong": a 422 is the user's
 * input to fix, a 401/403 is an operator configuration problem, and an unreachable
 * backend is neither.
 */

/** Mirrors BoundaryCategory in backend/app/schemas/boundary.py. */
export const BOUNDARY_CATEGORIES = [
  "WILDLIFE_SANCTUARY",
  "NATIONAL_PARK",
  "FOREST",
  "ECO_SENSITIVE_ZONE",
  "TIGER_RESERVE",
  "RAMSAR_WETLAND",
  "OTHER_RESTRICTED_ZONE",
] as const;

export type BoundaryCategory = (typeof BOUNDARY_CATEGORIES)[number];

export const CATEGORY_LABELS: Record<BoundaryCategory, string> = {
  WILDLIFE_SANCTUARY: "Wildlife Sanctuary",
  NATIONAL_PARK: "National Park",
  FOREST: "Forest",
  ECO_SENSITIVE_ZONE: "Eco-Sensitive Zone",
  TIGER_RESERVE: "Tiger Reserve",
  RAMSAR_WETLAND: "Ramsar Wetland",
  OTHER_RESTRICTED_ZONE: "Other Restricted Zone",
};

export type CollisionType = "DIRECT_COLLISION" | "BUFFER_COLLISION" | "NEARBY" | "CLEAR";
export type Severity = "LOW" | "MEDIUM" | "HIGH" | "CRITICAL";

export type GeoJsonGeometry = {
  type: "Polygon" | "MultiPolygon" | "Point";
  coordinates: unknown;
};

export type GeoJsonFeature = {
  type: "Feature";
  id?: string;
  geometry: GeoJsonGeometry | null;
  properties: Record<string, unknown> & { role?: "project_location" | "analysis_buffer" | "boundary" };
};

export type GeoJsonFeatureCollection = {
  type: "FeatureCollection";
  features: GeoJsonFeature[];
  notice?: string | null;
};

export type ProjectLocation = { latitude: number; longitude: number };

export type CollisionDetail = {
  boundary_id: string;
  boundary_name: string;
  category: BoundaryCategory;
  category_label: string;
  collision_type: CollisionType;
  severity: Severity;
  distance_meters: number;
  intersection_area_sqm: number;
  buffer_overlap_percentage: number;
  clearance_required: boolean;
  is_demo: boolean;
  state: string;
  district: string | null;
  geometry: GeoJsonGeometry | null;
};

export type ClearanceFlag = {
  boundary_id: string;
  boundary_name: string;
  category: BoundaryCategory;
  collision_type: CollisionType;
  severity: Severity;
  distance_meters: number;
  reason: string;
};

export type CollisionCheckResponse = {
  project_id: string | null;
  project_location: ProjectLocation;
  buffer_meters: number;
  proximity_threshold_meters: number;
  overall_status: CollisionType;
  overall_severity: Severity | null;
  clearance_required: boolean;
  boundaries_checked: number;
  candidates_examined: number;
  collision_count: number;
  collisions: CollisionDetail[];
  clearance_flags: ClearanceFlag[];
  geojson: GeoJsonFeatureCollection;
  projected_crs: string;
  contains_demo_data: boolean;
  notice: string | null;
  summary: string;
  analyzed_at: string;
};

export type CollisionCheckInput = {
  project_id?: string | null;
  latitude: number;
  longitude: number;
  buffer_meters: number;
  categories?: BoundaryCategory[] | null;
  include_demo?: boolean;
  include_geometry?: boolean;
};

export type NearbyBoundary = {
  boundary_id: string;
  name: string;
  category: BoundaryCategory;
  category_label: string;
  state: string;
  district: string | null;
  distance_meters: number;
  contains_point: boolean;
  severity: Severity;
  clearance_required: boolean;
  is_demo: boolean;
  geometry: GeoJsonGeometry | null;
};

export type NearbyBoundariesResponse = {
  project_location: ProjectLocation;
  radius_meters: number;
  categories: BoundaryCategory[] | null;
  count: number;
  boundaries: NearbyBoundary[];
  geojson: GeoJsonFeatureCollection;
  boundaries_checked: number;
  contains_demo_data: boolean;
  notice: string | null;
};

export type BoundarySummary = {
  id: string;
  name: string;
  category: BoundaryCategory;
  state: string;
  district: string | null;
  source: string;
  source_url: string | null;
  last_updated: string;
  area_sqkm: number;
  bbox: [number, number, number, number];
  is_demo: boolean;
};

export type BoundaryListResponse = {
  count: number;
  total: number;
  limit: number | null;
  offset: number;
  filters: Record<string, unknown>;
  boundaries: BoundarySummary[];
  geojson: GeoJsonFeatureCollection | null;
  contains_demo_data: boolean;
  notice: string | null;
};

export type BoundaryDetailResponse = {
  boundary: BoundarySummary;
  geometry: GeoJsonGeometry;
  feature: GeoJsonFeature;
  metadata: Record<string, unknown>;
  is_demo: boolean;
  notice: string | null;
};

export type IntersectingBoundary = Omit<CollisionDetail, "state" | "district" | "geometry">;

export type AssessmentRecord = {
  id: string;
  project_id: string;
  location: ProjectLocation;
  buffer_meters: number;
  created_at: string;
  created_by: string;
  created_by_name: string;
  overall_status: CollisionType;
  overall_severity: Severity | null;
  clearance_required: boolean;
  boundaries_checked: number;
  intersecting_boundaries: IntersectingBoundary[];
  clearance_flags: ClearanceFlag[];
  notes: string | null;
};

export type AssessmentHistoryResponse = {
  project_id: string;
  count: number;
  total: number;
  limit: number | null;
  offset: number;
  assessments: AssessmentRecord[];
};

// ---------------------------------------------------------------------------------
// Errors
// ---------------------------------------------------------------------------------

export class GisApiError extends Error {
  readonly status: number;
  constructor(message: string, status: number) {
    super(message);
    this.name = "GisApiError";
    this.status = status;
  }
}

/** 422 -- the submitted coordinates, buffer, or filters are not acceptable. */
export class GisValidationError extends GisApiError {
  readonly fields: string[];
  constructor(message: string, fields: string[]) {
    super(message, 422);
    this.name = "GisValidationError";
    this.fields = fields;
  }
}

/** 401 / 403 / 503-unconfigured -- an operator problem, not something the user typed. */
export class GisAuthError extends GisApiError {
  constructor(message: string, status: number) {
    super(message, status);
    this.name = "GisAuthError";
  }
}

/** 404 -- the boundary or project history does not exist. */
export class GisNotFoundError extends GisApiError {
  constructor(message: string) {
    super(message, 404);
    this.name = "GisNotFoundError";
  }
}

/** 5xx from the service itself. */
export class GisServerError extends GisApiError {
  constructor(message: string, status: number) {
    super(message, status);
    this.name = "GisServerError";
  }
}

/** The request never got an HTTP answer: offline, DNS, or the proxy could not connect. */
export class GisNetworkError extends GisApiError {
  constructor(message: string) {
    super(message, 0);
    this.name = "GisNetworkError";
  }
}

type FastApiValidationItem = { loc?: (string | number)[]; msg?: string };

/**
 * Turn a FastAPI error body into one readable sentence.
 *
 * FastAPI reports request-validation failures as an array of {loc, msg} objects and
 * everything else as a plain `detail` string, so both shapes have to be handled or the
 * user sees "[object Object]".
 */
function describeDetail(detail: unknown): { message: string; fields: string[] } {
  if (typeof detail === "string") return { message: detail, fields: [] };

  if (Array.isArray(detail)) {
    const fields: string[] = [];
    const parts = (detail as FastApiValidationItem[]).map((item) => {
      const path = (item.loc ?? []).filter((part) => part !== "body" && part !== "query");
      const field = path.join(".");
      if (field) fields.push(field);
      return field ? `${field}: ${item.msg ?? "invalid"}` : (item.msg ?? "invalid");
    });
    return { message: parts.join("; "), fields };
  }

  return { message: "The GIS service rejected the request.", fields: [] };
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  let response: Response;
  try {
    response = await fetch(`/api/gis/${path}`, {
      ...init,
      headers: { Accept: "application/json", ...(init?.headers ?? {}) },
    });
  } catch (error) {
    throw new GisNetworkError(
      `Could not reach the GIS service. Check your connection and that the backend is running. (${
        error instanceof Error ? error.message : "network error"
      })`,
    );
  }

  if (response.status === 204) return undefined as T;

  const payload = await response.json().catch(() => null);

  if (response.ok) return payload as T;

  const { message, fields } = describeDetail(payload?.detail);

  if (response.status === 422) throw new GisValidationError(message, fields);
  if (response.status === 401 || response.status === 403) throw new GisAuthError(message, response.status);
  if (response.status === 404) throw new GisNotFoundError(message);
  if (response.status === 503 && /authentication/i.test(message)) throw new GisAuthError(message, 503);
  // The proxy emits 502 for exactly one reason: it could not reach FastAPI at all. That is
  // a connectivity problem, not a fault in the service, so it belongs in the network state
  // alongside a failed fetch -- the user's next step is "start the backend", not "retry later".
  if (response.status === 502) throw new GisNetworkError(message);
  if (response.status >= 500) throw new GisServerError(message, response.status);
  throw new GisApiError(message, response.status);
}

function queryString(params: Record<string, string | number | boolean | string[] | null | undefined>): string {
  const search = new URLSearchParams();
  for (const [key, value] of Object.entries(params)) {
    if (value === null || value === undefined || value === "") continue;
    // Repeat the key for list filters -- the shape FastAPI expects for `list[...]` query params.
    if (Array.isArray(value)) value.forEach((item) => search.append(key, item));
    else search.append(key, String(value));
  }
  const encoded = search.toString();
  return encoded ? `?${encoded}` : "";
}

// ---------------------------------------------------------------------------------
// API surface
// ---------------------------------------------------------------------------------

export const gisApi = {
  /** POST /api/v1/gis/check-collision */
  checkCollision(input: CollisionCheckInput): Promise<CollisionCheckResponse> {
    return request<CollisionCheckResponse>("check-collision", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        project_id: input.project_id ?? null,
        latitude: input.latitude,
        longitude: input.longitude,
        buffer_meters: input.buffer_meters,
        categories: input.categories?.length ? input.categories : null,
        include_demo: input.include_demo ?? true,
        include_geometry: input.include_geometry ?? true,
      }),
    });
  },

  /** GET /api/v1/gis/nearby-boundaries */
  getNearbyBoundaries(params: {
    latitude: number;
    longitude: number;
    radius_meters: number;
    category?: BoundaryCategory[] | null;
    include_geometry?: boolean;
  }): Promise<NearbyBoundariesResponse> {
    return request<NearbyBoundariesResponse>(
      `nearby-boundaries${queryString({
        latitude: params.latitude,
        longitude: params.longitude,
        radius_meters: params.radius_meters,
        category: params.category ?? null,
        include_geometry: params.include_geometry ?? false,
      })}`,
    );
  },

  /** GET /api/v1/gis/boundaries */
  getBoundaries(params?: {
    category?: BoundaryCategory[] | null;
    state?: string | null;
    district?: string | null;
    search?: string | null;
    include_geometry?: boolean;
    limit?: number | null;
    offset?: number;
  }): Promise<BoundaryListResponse> {
    return request<BoundaryListResponse>(
      `boundaries${queryString({
        category: params?.category ?? null,
        state: params?.state ?? null,
        district: params?.district ?? null,
        search: params?.search ?? null,
        include_geometry: params?.include_geometry ?? false,
        limit: params?.limit ?? null,
        offset: params?.offset ?? 0,
      })}`,
    );
  },

  /** GET /api/v1/gis/boundaries/{id} */
  getBoundary(boundaryId: string): Promise<BoundaryDetailResponse> {
    return request<BoundaryDetailResponse>(`boundaries/${encodeURIComponent(boundaryId)}`);
  },

  /** GET /api/v1/gis/assessments/{project_id} -- 404 when the project has no history. */
  getAssessmentHistory(projectId: string, limit?: number): Promise<AssessmentHistoryResponse> {
    return request<AssessmentHistoryResponse>(
      `assessments/${encodeURIComponent(projectId)}${queryString({ limit: limit ?? null })}`,
    );
  },

  /** POST /api/v1/gis/assessments -- runs the engine and stores the outcome. */
  createAssessment(input: {
    project_id: string;
    latitude: number;
    longitude: number;
    buffer_meters: number;
    categories?: BoundaryCategory[] | null;
    notes?: string | null;
  }): Promise<AssessmentRecord> {
    return request<AssessmentRecord>("assessments", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        project_id: input.project_id,
        latitude: input.latitude,
        longitude: input.longitude,
        buffer_meters: input.buffer_meters,
        categories: input.categories?.length ? input.categories : null,
        notes: input.notes ?? null,
      }),
    });
  },
};

// ---------------------------------------------------------------------------------
// Display helpers -- formatting only; no analysis or classification happens client-side
// ---------------------------------------------------------------------------------

export const STATUS_LABELS: Record<CollisionType, string> = {
  DIRECT_COLLISION: "Direct collision",
  BUFFER_COLLISION: "Buffer collision",
  NEARBY: "Nearby boundary",
  CLEAR: "Clear",
};

export const STATUS_HEADLINES: Record<CollisionType, string> = {
  DIRECT_COLLISION: "Project location falls inside a restricted boundary",
  BUFFER_COLLISION: "Potential spatial conflict detected",
  NEARBY: "Restricted boundary close to the project location",
  CLEAR: "No spatial conflict detected",
};

/** Maps a severity onto the design system's existing pill classes. */
export const SEVERITY_CLASS: Record<Severity, string> = {
  CRITICAL: "critical",
  HIGH: "high",
  MEDIUM: "watch",
  LOW: "stable",
};

export function formatDistance(meters: number): string {
  if (meters === 0) return "Inside";
  if (meters < 1000) return `${Math.round(meters).toLocaleString()} m`;
  return `${(meters / 1000).toFixed(meters < 10_000 ? 2 : 1)} km`;
}

export function formatBuffer(meters: number): string {
  return meters >= 1000 ? `${(meters / 1000).toString().replace(/\.0$/, "")} km` : `${meters} m`;
}

export function formatArea(squareMeters: number): string {
  if (squareMeters <= 0) return "--";
  if (squareMeters < 1_000_000) return `${Math.round(squareMeters).toLocaleString()} m²`;
  return `${(squareMeters / 1_000_000).toFixed(2)} km²`;
}

File: backend/app/schemas/gis_api.py
"""Request/response contracts for the GIS REST API (Feature 6).

These are the HTTP-facing shapes only. They wrap -- and never reimplement -- the domain
models in app/schemas/spatial.py and app/schemas/boundary.py: every value a response
carries was computed by the spatial engine or read from the boundary store, and no field
here is populated by a default, a placeholder, or a hardcoded example.

Every response that can carry geometry carries it as RFC 7946 GeoJSON in WGS84, ready to
hand straight to Leaflet/MapLibre without transformation.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.config import spatial_config as config
from app.models.projection import validate_buffer_meters, validate_coordinates
from app.schemas.assessment import (
    AssessmentRecord,
    AssessmentSummary,
    ClearanceFlag,
    IntersectingBoundary,
    ProjectLocation,
)
from app.schemas.boundary import BoundaryCategory, BoundarySummary
from app.schemas.spatial import CollisionType, Severity

#: Role tags on the features of an analysis FeatureCollection, so a map can style the
#: project point, the tested buffer, and the boundaries differently without guessing.
FeatureRole = Literal["project_location", "analysis_buffer", "boundary", "collision_area"]


class ErrorResponse(BaseModel):
    """The error body every GIS endpoint returns. Documented so clients can rely on it."""

    detail: str = Field(description="A human-readable explanation of what went wrong and, where possible, how to fix it.")


class _CoordinateInput(BaseModel):
    """Shared coordinate/buffer validation for the request bodies below.

    Field bounds reject out-of-range values and NaN; the validators additionally route
    them through the engine's own checks, so HTTP and in-process callers can never
    disagree about what a valid coordinate is.
    """

    model_config = ConfigDict(extra="forbid")

    latitude: float = Field(ge=-90, le=90, description="WGS84 latitude in decimal degrees.")
    longitude: float = Field(ge=-180, le=180, description="WGS84 longitude in decimal degrees.")
    buffer_meters: float = Field(
        default=config.DEFAULT_BUFFER_METERS,
        ge=config.MIN_BUFFER_METERS,
        le=config.MAX_BUFFER_METERS,
        description="Buffer radius around the project point, in METRES.",
    )

    @field_validator("latitude")
    @classmethod
    def _valid_latitude(cls, value: float) -> float:
        validate_coordinates(value, 0.0)
        return value

    @field_validator("longitude")
    @classmethod
    def _valid_longitude(cls, value: float) -> float:
        validate_coordinates(0.0, value)
        return value

    @field_validator("buffer_meters")
    @classmethod
    def _valid_buffer(cls, value: float) -> float:
        return validate_buffer_meters(value)


# ---------------------------------------------------------------------------------
# API 1 -- collision check
# ---------------------------------------------------------------------------------


class CollisionCheckRequest(_CoordinateInput):
    """POST /api/v1/gis/check-collision"""

    project_id: str | None = Field(
        default=None, max_length=128, description="Optional. Echoed back; does not have to exist yet."
    )
    categories: list[BoundaryCategory] | None = Field(
        default=None, description="Optional. Restrict the scan to these boundary categories."
    )
    include_demo: bool = Field(default=True, description="Include fictional demo boundaries in the scan.")
    include_geometry: bool = Field(
        default=True, description="Attach each colliding boundary's GeoJSON geometry to the response."
    )


class CollisionDetail(IntersectingBoundary):
    """One colliding boundary, with the location context and geometry a map needs."""

    state: str
    district: str | None = None
    geometry: dict[str, Any] | None = Field(
        default=None, description="The boundary's GeoJSON Polygon/MultiPolygon in WGS84; null when omitted."
    )
    intersection_geometry: dict[str, Any] | None = Field(
        default=None,
        description=(
            "The exact buffer/boundary overlap as WGS84 GeoJSON, computed by the engine in a projected "
            "metre CRS. Null for NEARBY boundaries, which by definition do not overlap."
        ),
    )


class CollisionCheckResponse(BaseModel):
    """The engine's verdict for one project location."""

    project_id: str | None = None
    project_location: ProjectLocation
    buffer_meters: float
    proximity_threshold_meters: float

    overall_status: CollisionType = Field(description="Most severe collision type found; CLEAR when none.")
    overall_severity: Severity | None = Field(default=None, description="Highest severity found; null when CLEAR.")
    clearance_required: bool
    boundaries_checked: int = Field(ge=0, description="Boundaries in the dataset this scan ran against.")
    candidates_examined: int = Field(ge=0, description="Boundaries the spatial index selected for exact testing.")
    collision_count: int = Field(ge=0)

    collisions: list[CollisionDetail]
    clearance_flags: list[ClearanceFlag]

    geojson: dict[str, Any] = Field(
        description=(
            "A GeoJSON FeatureCollection with the project point, the tested buffer, every colliding "
            "boundary, and the exact overlap footprint of each intersecting one. Each feature's `role` "
            "property is one of project_location, analysis_buffer, boundary, collision_area."
        )
    )
    projected_crs: str = Field(description="The metre-based CRS all measurements were computed in.")
    contains_demo_data: bool
    notice: str | None = None
    summary: str
    analyzed_at: datetime


# ---------------------------------------------------------------------------------
# API 2 -- nearby boundaries
# ---------------------------------------------------------------------------------


class NearbyBoundary(BaseModel):
    """A boundary within the requested radius, with its measured distance."""

    boundary_id: str
    name: str
    category: BoundaryCategory
    category_label: str
    state: str
    district: str | None = None
    distance_meters: float = Field(ge=0, description="0 when the point lies inside this boundary.")
    contains_point: bool
    severity: Severity
    clearance_required: bool
    is_demo: bool
    geometry: dict[str, Any] | None = None


class NearbyBoundariesResponse(BaseModel):
    """GET /api/v1/gis/nearby-boundaries"""

    project_location: ProjectLocation
    radius_meters: float
    categories: list[BoundaryCategory] | None = None
    count: int = Field(ge=0)
    boundaries: list[NearbyBoundary]
    geojson: dict[str, Any]
    boundaries_checked: int = Field(ge=0)
    contains_demo_data: bool
    notice: str | None = None


# ---------------------------------------------------------------------------------
# API 3 / 4 -- boundary list and detail
# ---------------------------------------------------------------------------------


class BoundaryListResponse(BaseModel):
    """GET /api/v1/gis/boundaries"""

    count: int = Field(ge=0, description="Boundaries returned in this page.")
    total: int = Field(ge=0, description="Boundaries matching the filters, before limit/offset.")
    limit: int | None = None
    offset: int = Field(ge=0)
    filters: dict[str, Any] = Field(description="The filters actually applied, echoed for clients that cache.")
    boundaries: list[BoundarySummary]
    geojson: dict[str, Any] | None = Field(
        default=None, description="A FeatureCollection of the returned boundaries; null unless requested."
    )
    contains_demo_data: bool
    notice: str | None = None


class BoundaryDetailResponse(BaseModel):
    """GET /api/v1/gis/boundaries/{id}"""

    boundary: BoundarySummary
    geometry: dict[str, Any] = Field(description="The boundary's GeoJSON Polygon/MultiPolygon in WGS84.")
    feature: dict[str, Any] = Field(description="The same record as a complete GeoJSON Feature.")
    metadata: dict[str, Any]
    is_demo: bool
    notice: str | None = None


class BoundaryCreateRequest(BaseModel):
    """POST /api/v1/gis/boundaries -- requires MANAGE_GIS_DATA."""

    model_config = ConfigDict(extra="forbid")

    name: str = Field(min_length=1, max_length=300)
    category: BoundaryCategory
    state: str = Field(min_length=1, max_length=120)
    district: str | None = Field(default=None, max_length=120)
    geometry: dict[str, Any] = Field(description="GeoJSON Polygon or MultiPolygon in WGS84.")
    source: str = Field(min_length=1, max_length=300, description="Required: no boundary may exist without an origin.")
    source_url: str | None = Field(default=None, max_length=1000)
    last_updated: str = Field(description="ISO date the SOURCE dataset was last updated.")
    metadata: dict[str, Any] = Field(default_factory=dict)
    is_demo: bool = Field(
        default=True,
        description=(
            "Defaults to true. Set false ONLY for a genuinely official or licensed dataset -- "
            "records marked official are presented without the demo-data notice."
        ),
    )
    overwrite: bool = Field(default=False, description="Replace an existing record with the same derived id.")


# ---------------------------------------------------------------------------------
# API 5 -- assessments
# ---------------------------------------------------------------------------------


class AssessmentCreateRequest(_CoordinateInput):
    """POST /api/v1/gis/assessments -- runs the engine and stores the outcome."""

    project_id: str = Field(min_length=1, max_length=128, description="Required: an assessment belongs to a project.")
    categories: list[BoundaryCategory] | None = None
    include_demo: bool = True
    notes: str | None = Field(default=None, max_length=2000)


class AssessmentHistoryResponse(BaseModel):
    """GET /api/v1/gis/assessments/{project_id}"""

    project_id: str
    count: int = Field(ge=0)
    total: int = Field(ge=0)
    limit: int | None = None
    offset: int = Field(ge=0)
    assessments: list[AssessmentRecord]


class AssessmentHistorySummaryResponse(BaseModel):
    """The same history without full result payloads."""

    project_id: str
    count: int = Field(ge=0)
    total: int = Field(ge=0)
    assessments: list[AssessmentSummary]
